Use the in-tree `compiler-builtins` for the sysroot

tgross35 opened the pull request:
Many of `std`'s dependency have a dependency on the crates.io `compiler-builtins` when used with the feature `rustc-std-workspace-core`. Use a Cargo patch to select the in-tree version instead.

`compiler-builtins` is also added as a dependency of `rustc-std-workspace-core` so these crates can remove their crates.io dependency in the future.

Zulip discussion: [#t-compiler > Using in-tree compiler-builtins](https://rust-lang.zulipchat.com/#narrow/channel/131828-t-compiler/topic/Using.20in-tree.20compiler-builtins/with/522445336)

Once this merges, the following PRs will need to make it to a release for the relevant crates:

- https://github.com/rust-lang/getopts/pull/119 (can merge at any time) released in v0.2.23
- https://github.com/rust-lang/hashbrown/pull/625 (can merge at any time) released in v0.15.4
- https://github.com/rust-lang/stdarch/pull/1825
- https://github.com/rust-lang/rustc-demangle/pull/80
- https://github.com/rust-lang/cfg-if/pull/84
- https://github.com/unicode-rs/unicode-width/pull/77

The above should cover all tier 1 targets with no `std` features enabled. The remaining cover the rest:

- https://github.com/alexcrichton/dlmalloc-rs/pull/50 (wasm, xous, sgx)
- https://github.com/gimli-rs/gimli/pull/769
- https://github.com/r-efi/r-efi/pull/89 (efi)
- https://github.com/r-efi/r-efi-alloc/pull/9 (efi)
- https://github.com/fortanix/rust-sgx/pull/770 (sgx)
- https://github.com/hermit-os/hermit-rs/pull/718 (hermit)
- https://github.com/bytecodealliance/wasi-rs/pull/108 (wasi)
- https://github.com/gimli-rs/addr2line/pull/345
- https://github.com/oyvindln/adler2/pull/2
- https://github.com/BurntSushi/memchr/pull/180
- https://github.com/Frommi/miniz_oxide/pull/173
- https://github.com/gimli-rs/object/pull/777

try-job: x86_64-gnu
try-job: test-various

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-19`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/15430568241/job/43427514055) [(plain)](https://github.com/rust-lang/rust/commit/833fe4adfcf8e245affe1a336d96268082b64ed2/checks/43427514055/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#19 exporting to docker image format
#19 sending tarball 26.8s done
#19 DONE 34.2s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-19]
[CI_JOB_NAME=x86_64-gnu-llvm-19]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Listening on address 127.0.0.1:4226
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-19', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'build.print-step-timings', '--enable-verbose-tests', '--set', 'build.metrics', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--set', 'gcc.download-ci-gcc=true', '--enable-new-symbol-mangling']
configure: build.build          := x86_64-unknown-linux-gnu
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-19/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
```

</details>

bjorn3 commented on compiler/rustc_metadata/src/creader.rs:
```diff
@@ -1195,9 +1195,8 @@ impl<'a, 'tcx> CrateLoader<'a, 'tcx> {
         }
 
         // `compiler_builtins` is not yet in the graph; inject it. Error on resolution failure.
```
```suggestion
        // `compiler_builtins` is not yet in the graph; inject it.
```

bjorn3 replied:
Linker failures are less user friendly that rustc errors. Also it would allow accidentally depending on rustc not emitting any compiler-builtins calls for a given crate and thus linking succeeding with one rustc version and then failing on the next due to an optimization change.

tgross35 replied:
Makes sense. I wonder if it would make sense to try the sysroot if it can't resolve, but that doesn't come up very much.

bjorn3 replied:
This should already try resolving from the sysroot if there is no `--extern compiler_builtins=...`.

tgross35 commented:
@bors2 try

tgross35 committed:
Use the in-tree `compiler-builtins`

Many of `std`'s dependency have a dependency on the crates.io
`compiler-builtins` when used with the feature
`rustc-std-workspace-core`. Use a Cargo patch to select the in-tree
version instead.

`compiler-builtins` is also added as a dependency of
`rustc-std-workspace-core` so these crates can remove their crates.io
dependency in the future.

tgross35 commented:
@bors2 cancel

tgross35 commented:
@bors2 try cancel

tgross35 commented on library/compiler-builtins/compiler-builtins/Cargo.toml:
```diff
@@ -1,15 +1,15 @@
 edition = "2021"
 description = "Compiler intrinsics used by the Rust compiler."
-links = "compiler-rt"
+# links = "compiler-rt"
```
So something like the current CI failure is why I had this

```
Building stage2 miri sysroot (x86_64-unknown-linux-gnu)
  Preparing a sysroot for Miri (target: x86_64-unknown-linux-gnu)...
      Updating crates.io index
  error: failed to select a version for `compiler_builtins`.
      ... required by package `r-efi-alloc v2.0.0`
      ... which satisfies dependency `r-efi-alloc = "^2.0.0"` (locked to 2.0.0) of package `std v0.0.0 (/checkout/library/std)`
      ... which satisfies path dependency `std` (locked to 0.0.0) of package `custom-local-sysroot v0.0.0 (/tmp/.tmp1x07Mk)`
  versions that meet the requirements `^0.1.79` are: 0.1.160, 0.1.159, 0.1.158, 0.1.157, 0.1.156, 0.1.155, 0.1.154, 0.1.153, 0.1.152, 0.1.151, 0.1.150, 0.1.149, 0.1.148, 0.1.147, 0.1.146, 0.1.145, 0.1.144, 0.1.143, 0.1.142, 0.1.141, 0.1.140, 0.1.139, 0.1.138, 0.1.137, 0.1.136, 0.1.135, 0.1.134, 0.1.133, 0.1.132, 0.1.131, 0.1.130, 0.1.129, 0.1.128, 0.1.127, 0.1.126, 0.1.125, 0.1.124, 0.1.123, 0.1.122, 0.1.121, 0.1.120, 0.1.119, 0.1.118, 0.1.117, 0.1.116, 0.1.115, 0.1.114, 0.1.113, 0.1.112, 0.1.111, 0.1.110, 0.1.109, 0.1.108, 0.1.107, 0.1.106, 0.1.105, 0.1.104, 0.1.103, 0.1.102, 0.1.101, 0.1.100, 0.1.99, 0.1.98, 0.1.97, 0.1.96, 0.1.95, 0.1.94, 0.1.93, 0.1.92, 0.1.91, 0.1.90, 0.1.89, 0.1.88, 0.1.87, 0.1.86, 0.1.85, 0.1.84, 0.1.83, 0.1.82, 0.1.81, 0.1.80, 0.1.79
  
  package `compiler_builtins` links to the native library `compiler-rt`, but it conflicts with a previous package which links to `compiler-rt` as well:
  package `compiler_builtins v0.1.160 (/checkout/library/compiler-builtins/compiler-builtins)`
      ... which satisfies path dependency `compiler_builtins` (locked to 0.1.160) of package `std v0.0.0 (/checkout/library/std)`
      ... which satisfies path dependency `std` (locked to 0.0.0) of package `custom-local-sysroot v0.0.0 (/tmp/.tmp1x07Mk)`
  Only one package in the dependency graph may specify the same links value. This helps ensure that only one copy of a native library is linked in the final binary. Try to adjust your dependencies so that only one package uses the `links = "compiler-rt"` value. For more information, see https://doc.rust-lang.org/cargo/reference/resolver.html#links.
  
  failed to select a version for `compiler_builtins` which could resolve this conflict
  fatal error: failed to build sysroot: sysroot build failed
```

I have no idea where the other version could be coming from

tgross35 replied:
Resolved again.

tgross35 commented:
@bors2 try

tgross35 commented:
@bors2 try cancel

tgross35 commented:
I believe this works now. The first two commits are cherry picks that I will drop once https://github.com/rust-lang/rust/pull/142036 merges.

r? @bjorn3

tgross35 commented:
@bors2 try

tgross35 committed:
compiler-builtins: Fix a `rustdoc::bare-urls` error

RalfJung commented on src/tools/miri/cargo-miri/src/setup.rs:
```diff
@@ -128,6 +128,12 @@ pub fn setup(
             }
         }
 
+        command.arg(&format!(
+            "--config=patch.crates-io.compiler_builtins.path=\
+            '{}/compiler-builtins/compiler-builtins'",
+            rust_src.display(),
+        ));
```
cargo's `-Zbuild-std` will likely also need some sort of patch to keep working?

Will this affect RA's ability to load the sysroot?

RalfJung replied:
To be clear, having this here in Miri is IMO not something we should land, even if we go with the "temporary patch" solution. This needs to happen in https://github.com/RalfJung/rustc-build-sysroot/ and it needs to be done in a way that it works with old and new sysroots.

tgross35 replied:
Understood - I'll prep a PR to that as well as a handful of the compiler-builtins reverse dependencies once we're more certain about the path forward here.

tgross35 commented:
> Do they need that dependency? If not, can we remove it rather than adding these patches (which seems like it could very easily cause issues)?

Before this PR, anything that is `#![no_std]` and doesn't depend on `alloc` does need to specify `compiler_builtins` in their graph, so they use it from crates.io. After this PR they shouldn't need it since they'll get it automatically from `rustc_std_workspace_core` - after which point we can go around and delete `compiler_builtins` from their Cargo.toml.

So using the patch should only be needed in the interim time because having both the in-tree and the crates.io version in the graph makes things go haywire.

RalfJung commented:
That sounds like we should first transition to having the dependency via rustc-std-workspace-core, and then swap that over to an in-tree dependency, without ever needing such patches. It'll take two PRs instead of one, but it will avoid a fragile and painful intermediate state that everyone building sysroots will have to find manual work-arounds for.

RalfJung commented:
The alternative would be to
- ping the cargo build-std people to figure out whether they need any adjustments
- land a PR in https://github.com/RalfJung/rustc-build-sysroot that ensures building the sysroot works before and after the patch (i.e., adding the `[patch]` section only if needed)
- being okay with breaking xargo (it is unmaintained so it may well be broken already)

RalfJung commented:
Sounds like we need a `rustc-std-workspace-compiler-builtins` that published crates can depend on so that we can control what that dependency actually becomes?

tgross35 commented:
That was an alternative - but then updates to the patch section are definitely needed :) By instead making it a dependency of `rustc-std-workspace-core` at least crates that use it get to clean up a dependency rather than replacing it, and presumably something like `xargo` would automatically go back to working after the new patch section is no longer needed.

@bors2 try

tgross35 commented:
Test with https://github.com/RalfJung/rustc-build-sysroot/pull/27

@bors2 try

tgross35 commented:
> @rust-lang/cargo @rust-lang/rust-analyzer does anyone have a rough idea of what changes, if any, would be needed with this PR to keep from breaking `-Zbuild-std` and R-A's building of the sysroot? The relevant part is that a new (hopefully short lived) patch item in `library/Cargo.toml` would be required.

@ehuss was able to verify this didn't seem to break Cargo things: https://rust-lang.zulipchat.com/#narrow/channel/246057-t-cargo/topic/Testing.20with.20a.20specific.20rust-lang.2Frust.20PR/with/522678293

RalfJung commented:
Also, please change miri's `Cargo.toml` to require `0.5.8` so that on the next sync to the Miri repo it'll be clear what we have to do.

bjorn3 commented on compiler/rustc_codegen_gcc/build_system/build_sysroot/Cargo.toml:
```diff
@@ -16,6 +15,7 @@ proc_macro = { path = "./sysroot_src/library/proc_macro" }
 rustc-std-workspace-core = { path = "./sysroot_src/library/rustc-std-workspace-core" }
 rustc-std-workspace-alloc = { path = "./sysroot_src/library/rustc-std-workspace-alloc" }
 rustc-std-workspace-std = { path = "./sysroot_src/library/rustc-std-workspace-std" }
+compiler_builtins = { path = "./sysroot_src/library/compiler-builtins/compiler-builtins" }
```
@antoyo It should be possible to remove `build_sysroot/Cargo.toml` and use `sysroot_src/library/sysroot/Cargo.toml` instead. That is what cg_clif does nowadays too.

bjorn3 commented on src/tools/miri/cargo-miri/Cargo.toml:
```diff
@@ -18,7 +18,7 @@ directories = "6"
 serde_json = "1.0.40"
 cargo_metadata = "0.19"
-rustc-build-sysroot = "0.5.7"
+rustc-build-sysroot = "0.5.8"
```
Please update `cargo-miri/Cargo.lock`.

tgross35 replied:
Updated, thanks.

bjorn3 reviewed (approved):
r=me with the cargo-miri lockfile updated.

tgross35 commented:
@bors r=bjorn3

bors commented:
:pushpin: Commit f43bb2ac994e356d6e3abbee6f4eb1609a7bd0bd has been approved by `bjorn3`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=bjorn3 f43bb2ac994e356d6e3abbee6f4eb1609a7bd0bd -->
<!-- homu: {"type":"Approved","sha":"f43bb2ac994e356d6e3abbee6f4eb1609a7bd0bd","approver":"bjorn3","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

tgross35 committed:
compiler-builtins: Resolve `unsafe_op_in_unsafe_fn` on Arm32 Android

There are a few places that violate this lint, which showed up in
rust-lang/rust CI (the relevent module is gated behind
`kernel_user_helpers` which is only set for `armv4t`, `armv5te`, and
`arm-linux-androideabi`; none of these are tested in compiler-builtins
CI). Add new `unsafe { /* ... */ }` blocks where needed to address this.

Some blocks should get a more thorough review of their preconditions, so
their safety comments are left as `FIXME`s.

bjorn3 commented:
@bors r+

bors commented:
:hourglass: Testing commit 974d43733225774a449a936de354f3b1cd29a85a with merge 8d6c357c871e5b7daaaaf64e1b3b51683697a880...
<!-- homu: {"type":"BuildStarted","head_sha":"974d43733225774a449a936de354f3b1cd29a85a","merge_sha":"8d6c357c871e5b7daaaaf64e1b3b51683697a880"} -->

rust-log-analyzer commented:
The job **`dist-various-2`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/15519418003/job/43690668746) [(plain)](https://github.com/rust-lang/rust/commit/8d6c357c871e5b7daaaaf64e1b3b51683697a880/checks/43690668746/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
   Compiling std v0.0.0 (/checkout/library/std)
error: formatting may not be suitable for sub-register argument
  --> library/compiler-builtins/compiler-builtins/src/mem/x86_64.rs:72:14
   |
72 |         "mov {qword_count}, %rcx",
   |              ^^^^^^^^^^^^^
...
81 |         qword_count = in(reg) qword_count,
   |                               ----------- for this argument
   |
   = help: use `{1:e}` to have the register formatted as `eax` (for 32-bit values)
   = help: or use `{1:r}` to keep the default formatting of `rax` (for 64-bit values)
   = note: `-D asm-sub-register` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(asm_sub_register)]`

error: formatting may not be suitable for sub-register argument
   --> library/compiler-builtins/compiler-builtins/src/mem/x86_64.rs:215:22
    |
215 |             "movdqa ({addr}), {dest}",
    |                      ^^^^^^
216 |             addr = in(reg) s,
    |                            - for this argument
    |
    = help: use `{0:e}` to have the register formatted as `eax` (for 32-bit values)
    = help: or use `{0:r}` to keep the default formatting of `rax` (for 64-bit values)

error: formatting may not be suitable for sub-register argument
   --> library/compiler-builtins/compiler-builtins/src/mem/x86_64.rs:235:26
    |
235 |                 "movdqa ({addr}), {dest}",
    |                          ^^^^^^
236 |                 addr = in(reg) s,
    |                                - for this argument
    |
    = help: use `{0:e}` to have the register formatted as `eax` (for 32-bit values)
    = help: or use `{0:r}` to keep the default formatting of `rax` (for 64-bit values)

[RUSTC-TIMING] compiler_builtins test:false 1.352
error: could not compile `compiler_builtins` (lib) due to 3 previous errors
warning: build failed, waiting for other jobs to finish...
[RUSTC-TIMING] core test:false 21.081
```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang/rust/actions/runs/15519418003/job/43691760859)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang/rust/actions/runs/15519418003/job/43691760859","builder_name":"checks-actions"} -->

tgross35 committed:
compiler-builtins: Specify `:r` registers for `usize`

On the ILP32 `x86_64-unknown-linux-gnux32` target, `usize` is 32 bits so
there is a sub-register alignment warning. Specify the 64-bit `r`
registers, which matches the current default as well as the size of the
other operands in the routines.

tgross35 commented:
@bors2 try jobs=`*various*`

tgross35 committed:
compiler-builtins: Emit `rustc-check-cfg` earlier

The `build.rs` entrypoint returns early for some targets, so emscripten
and OpenBSD were not getting check-cfg set. Emit these earlier to avoid
the `unexpected_cfgs` lint.

tgross35 commented:
Two small lint-related fixes on targets that aren't tested in compiler-builtins CI. I also locally checked that compiler-builtins builds without warning on all of `rustc --print target-list`.

@bors r=bjorn3

bors commented:
:pushpin: Commit fc715449326ce5dc193bc717c854c8d35c07508e has been approved by `bjorn3`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=bjorn3 fc715449326ce5dc193bc717c854c8d35c07508e -->
<!-- homu: {"type":"Approved","sha":"fc715449326ce5dc193bc717c854c8d35c07508e","approver":"bjorn3","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
